File: casper/src/main/scala/coop/rchain/casper/util/MergeableChannelsGC.scala
package coop.rchain.casper.util

import cats.effect.Sync
import cats.syntax.all._
import coop.rchain.blockstorage.{BlockStore, BlockStoreSyntax}
import coop.rchain.blockstorage.dag.BlockDagRepresentation
import coop.rchain.blockstorage.dag.BlockDagRepresentationSyntax
import coop.rchain.casper.CasperShardConf
import coop.rchain.casper.PrettyPrinter
import coop.rchain.casper.util.rholang.RuntimeManager
import coop.rchain.metrics.Metrics
import coop.rchain.models.BlockHash.BlockHash
import coop.rchain.models.syntax._
import coop.rchain.shared.Log
import coop.rchain.store.KeyValueTypedStoreSyntax

object MergeableChannelsGC
    extends BlockDagRepresentationSyntax
    with BlockStoreSyntax
    with KeyValueTypedStoreSyntax {

  /**
    * Garbage collects mergeable channel data for blocks that are provably unreachable.
    *
    * A block's mergeable data is safe to delete when:
    * 1. The block is finalized
    * 2. All validators' latest messages are descendants of the block's children
    * 3. The block is deeper than maxParentDepth + depthBuffer from current tips
    */
  def collectGarbage[F[_]: Sync: Log: Metrics: BlockStore](
      dag: BlockDagRepresentation[F],
      runtimeManager: RuntimeManager[F],
      casperShardConf: CasperShardConf
  )(implicit ms: Metrics.Source): F[Unit] = {

    def isSafeToDelete(blockHash: BlockHash): F[Boolean] =
      for {
        blockMeta <- dag.lookupUnsafe(blockHash)

        // Check if block is finalized
        isFinalized <- dag.isFinalized(blockHash)

        result <- if (!isFinalized) {
                   false.pure[F]
                 } else {
                   for {
                     // Check depth constraint
                     maxBlockNumber <- dag.latestBlockNumber
                     depthFromTip   = maxBlockNumber - blockMeta.blockNum
                     maxAllowedDepth = casperShardConf.maxParentDepth +
                       casperShardConf.mergeableChannelsGCDepthBuffer
                     tooDeep = depthFromTip > maxAllowedDepth

                     allMovedPast <- if (!tooDeep) {
                                      false.pure[F]
                                    } else {
                                      // Check if all validators have moved past this block
                                      for {
                                        latestMessagesHashes <- dag.latestMessageHashes
                                        childrenOpt          <- dag.children(blockHash)
                                        children             = childrenOpt.getOrElse(Set.empty)

                                        // For each validator's latest message, check if it's a descendant
                                        // of at least one child of this block
                                        allMovedPast <- if (children.isEmpty) {
                                                         // No children means no one can have moved past
                                                         false.pure[F]
                                                       } else {
                                                         latestMessagesHashes.values.toList
                                                           .traverse { latestMsgHash =>
                                                             if (latestMsgHash == blockHash) {
                                                               // Validator's latest is still this block
                                                               false.pure[F]
                                                             } else {
                                                               // Check if latest message is descendant of any child
                                                               children.toList.existsM {
                                                                 childHash =>
                                                                   dag.isInMainChain(
                                                                     childHash,
                                                                     latestMsgHash
                                                                   )
                                                               }
                                                             }
                                                           }
                                                           .map(_.forall(identity))
                                                       }
                                      } yield allMovedPast
                                    }
                   } yield tooDeep && allMovedPast
                 }
      } yield result

    for {
      // Get all finalized blocks
      finalizedBlocks <- dag
                          .topoSort(0, None)
                          .map(_.flatten)
                          .flatMap(_.filterA(dag.isFinalized))

      // Find candidates that are safe to delete
      candidates <- finalizedBlocks.filterA(isSafeToDelete)

      // Delete mergeable data for safe candidates
      _ <- candidates.traverse { blockHash =>
            for {
              block                             <- BlockStore[F].getUnsafe(blockHash)
              stateHash: scodec.bits.ByteVector = block.body.state.postStateHash.toBlake2b256Hash.bytes
              _                                 <- runtimeManager.getMergeableStore.delete(stateHash)
              _ <- Log[F].debug(
                    s"GC: Deleted mergeable data for block ${PrettyPrinter.buildString(blockHash)}"
                  )
            } yield ()
          }

      // Log and record metrics
      _ <- if (candidates.nonEmpty) {
            Metrics[F].incrementCounter("mergeable_channels_gc_deleted", candidates.size.toLong) >>
              Log[F].info(s"Mergeable channels GC: Deleted ${candidates.size} blocks' data")
          } else {
            Log[F].debug("Mergeable channels GC: No data to delete")
          }
    } yield ()
  }
}

File: casper/src/rust/util/mergeable_channels_gc.rs
//! Mergeable Channels Garbage Collection
//!
//! Garbage collects mergeable channel data for blocks that are provably unreachable.
//! This is required for multi-parent mode where immediate deletion during finalization
//! can cause data races.

use block_storage::rust::dag::block_dag_key_value_storage::KeyValueDagRepresentation;
use block_storage::rust::key_value_block_store::KeyValueBlockStore;
use models::rust::block_hash::BlockHash;
use shared::rust::store::key_value_store::KvStoreError;

use crate::rust::casper::CasperShardConf;
use crate::rust::util::rholang::runtime_manager::RuntimeManager;

/// Garbage collects mergeable channel data for blocks that are provably unreachable.
///
/// A block's mergeable data is safe to delete when:
/// 1. The block is finalized
/// 2. All validators' latest messages are descendants of the block's children
/// 3. The block is deeper than maxParentDepth + depthBuffer from current tips
pub async fn collect_garbage(
    dag: &KeyValueDagRepresentation,
    block_store: &KeyValueBlockStore,
    runtime_manager: &std::sync::Arc<tokio::sync::Mutex<RuntimeManager>>,
    casper_shard_conf: &CasperShardConf,
) -> Result<usize, KvStoreError> {
    let mut deleted_count = 0;

    // Get all finalized blocks by traversing from genesis
    // Note: This could be optimized by tracking pending GC blocks
    let finalized_blocks = get_finalized_blocks(dag)?;

    for block_hash in finalized_blocks {
        if is_safe_to_delete(dag, &block_hash, casper_shard_conf)? {
            // Get block to access its state hash
            if let Some(block) = block_store.get(&block_hash)? {
                let deleted = runtime_manager
                    .lock()
                    .await
                    .delete_mergeable_channels(
                        &block.body.state.post_state_hash,
                        block.sender.clone(),
                        block.seq_num,
                    )
                    .map_err(|e| KvStoreError::IoError(e.to_string()))?;

                if deleted {
                    deleted_count += 1;
                    tracing::debug!(
                        "GC: Deleted mergeable data for block {}",
                        hex::encode(&block_hash)
                    );
                }
            }
        }
    }

    if deleted_count > 0 {
        metrics::counter!("mergeable_channels_gc_deleted").increment(deleted_count as u64);
        tracing::info!(
            "Mergeable channels GC: Deleted {} blocks' data",
            deleted_count
        );
    } else {
        tracing::debug!("Mergeable channels GC: No data to delete");
    }

    Ok(deleted_count)
}

/// Check if a block's mergeable data is safe to delete.
fn is_safe_to_delete(
    dag: &KeyValueDagRepresentation,
    block_hash: &BlockHash,
    casper_shard_conf: &CasperShardConf,
) -> Result<bool, KvStoreError> {
    // 1. Check if block is finalized
    if !dag.is_finalized(block_hash) {
        return Ok(false);
    }

    // 2. Check depth constraint
    let block_meta = dag.lookup_unsafe(block_hash)?;
    let max_block_number = dag.latest_block_number();
    let depth_from_tip = max_block_number - block_meta.block_number;
    let max_allowed_depth = (casper_shard_conf.max_parent_depth as i64)
        + (casper_shard_conf.mergeable_channels_gc_depth_buffer as i64);

    if depth_from_tip <= max_allowed_depth {
        return Ok(false);
    }

    // 3. Check if all validators have moved past this block
    let children = match dag.children(block_hash) {
        Some(children_set) => children_set,
        None => return Ok(false), // No children means no one can have moved past
    };

    if children.is_empty() {
        return Ok(false);
    }

    let latest_message_hashes = dag.latest_message_hashes();

    // For each validator's latest message, check if it's a descendant of any child (via main chain)
    for (_, latest_msg_hash) in latest_message_hashes.iter() {
        if latest_msg_hash == block_hash {
            // Validator's latest is still this block
            return Ok(false);
        }

        // Check if latest message is descendant of any child (via main chain)
        let mut found_in_child_chain = false;
        for child_hash_ref in children.iter() {
            if dag.is_in_main_chain(child_hash_ref, latest_msg_hash)? {
                found_in_child_chain = true;
                break;
            }
        }

        if !found_in_child_chain {
            return Ok(false);
        }
    }

    Ok(true)
}

/// Get all finalized blocks from the DAG.
/// Note: This is a simple implementation that could be optimized.
fn get_finalized_blocks(dag: &KeyValueDagRepresentation) -> Result<Vec<BlockHash>, KvStoreError> {
    // Get all blocks via topo_sort and filter for finalized ones
    let all_blocks = dag.topo_sort(0, None)?;

    let finalized: Vec<BlockHash> = all_blocks
        .into_iter()
        .flatten()
        .filter(|hash| dag.is_finalized(hash))
        .collect();

    Ok(finalized)
}

#[cfg(test)]
mod tests {
    // Tests would go here
}
